### PR TITLE
Issue#101 Old User Data accessible

### PR DESCRIPTION
**Pull request by dayitv89**
User auth token stored in the keychain, when app deleted and reinstall this data won’t remove and previous user auth token accessed, if it expires it auto refresh the auth token.

Fixes: Store a bool value to user default and check it is first run then remove auth token else previous behaviour.

### CODE REVIEW

**Comment by jbrophy17** [comment]
Hi @dayitv89 thanks for the PR! Before I can merge this, could you add a few tests to verify that this functionality works as you intend? Thanks!

**Comment by dayitv89** [comment]
@jbrophy17 Have a look for the test cases. Sorry for the delay.

**Comment by jbrophy17** [comment]
Thanks for adding some tests! I will take a look at this and let you know any feedback

File: source/UberRides/TokenManager.swift [hunk 1]
```diff
@@ -143,36 +150,38 @@ import Foundation
         deleteCookies()
         let success = keychainWrapper.deleteObjectForKey(tokenIdentifier)
         if success {
+            NSUserDefaults.standardUserDefaults().removeObjectForKey(kCheckFirstRun)
```
Comment (jbrophy17) [inline source/UberRides/TokenManager.swift]: If we delete the first run key whenever you delete a key, the next key you fetch is going to be nil.

For example, we could save 2 keys with 2 separate identifiers and then delete one of them. When we try to access the second key, the `fetch()` check will think it's a first run and delete the key

File: source/UberRidesTests/RidesClientTests.swift [hunk 2]
```diff
@@ -1083,4 +1087,11 @@ class RidesClientTests: XCTestCase {
         let accessToken = ridesClient.fetchAccessToken()
         XCTAssertNil(accessToken)
     }
+    
+    //MARK: Helper
+    func saveFirstRunKey() {
+        let kCheckFirstRun = "com.uber.checkFirstRun"
```
Comment (jbrophy17) [inline source/UberRidesTests/RidesClientTests.swift]: let's make `kCheckFirstRun` internal in TokenManager and use that value here rather than declaring it again

File: source/UberRidesTests/RidesClientTests.swift [hunk 3]
```diff
@@ -1083,4 +1087,11 @@ class RidesClientTests: XCTestCase {
         let accessToken = ridesClient.fetchAccessToken()
         XCTAssertNil(accessToken)
     }
+    
+    //MARK: Helper
+    func saveFirstRunKey() {
```
Comment (jbrophy17) [inline source/UberRidesTests/RidesClientTests.swift]: I think it would make sense to move this whole function to just be part of the setUp() of the test case. With corresponding logic in tearDown() to remove the key

File: source/UberRidesTests/TokenManagerTests.swift [hunk 4]
```diff
@@ -26,210 +26,236 @@ import XCTest
 @testable import UberRides
 
 class TokenManagerTests: XCTestCase {
-    
+
     private var notificationFired = false
     private var keychain: KeychainWrapper?
-    
+    private let kCheckFirstRun = "com.uber.checkFirstRun"
```
Comment (jbrophy17) [inline source/UberRidesTests/TokenManagerTests.swift]: same here

File: source/UberRidesTests/TokenManagerTests.swift [hunk 5]
```diff
@@ -26,210 +26,236 @@ import XCTest
...
+        XCTAssertNotNil(actualToken)
+        XCTAssertEqual(actualToken?.tokenString, token.tokenString)
+
+        // When app deleted, cleared the user default
+        NSUserDefaults.standardUserDefaults().removeObjectForKey(kCheckFirstRun)
+        NSUserDefaults.standardUserDefaults().synchronize()
+
+        let deletedToken = TokenManager.fetchToken(identifier, accessGroup: accessGroup)
+        XCTAssertNil(deletedToken)
+        XCTAssertNotEqual(deletedToken?.tokenString, token.tokenString)
```
Comment (jbrophy17) [inline source/UberRidesTests/TokenManagerTests.swift]: I don't think we need this since we already test for `nil`

File: source/UberRidesTests/TokenManagerTests.swift [hunk 6]
```diff
@@ -26,210 +26,236 @@ import XCTest
...
+
         let testCookieStorage = NSHTTPCookieStorage.sharedHTTPCookieStorage()
         XCTAssertEqual(testCookieStorage.cookies?.count, 0)
-        
+
     }
-    
-    
+
+    func testAppDeleteFunctionality() {
```
Comment (jbrophy17) [inline source/UberRidesTests/TokenManagerTests.swift]: Can we rename this to `testTokenDeletedOnFirstRun()` or something similar?

File: source/UberRidesTests/TokenManagerTests.swift [hunk 7]
```diff
@@ -26,210 +26,236 @@ import XCTest
...
+        NSUserDefaults.standardUserDefaults().synchronize()
+
+        let deletedToken = TokenManager.fetchToken(identifier, accessGroup: accessGroup)
+        XCTAssertNil(deletedToken)
+        XCTAssertNotEqual(deletedToken?.tokenString, token.tokenString)
+
+        TokenManager.deleteToken(identifier, accessGroup: accessGroup)
+    }
+
+
```
Comment (jbrophy17) [inline source/UberRidesTests/TokenManagerTests.swift]: Can we add a test case for the situation I commented above? Basically:

save token with identifier-a
save token with identifier-b
verify both not nil
delete token with identifier-a
verify token with identifier-a is nil
verify token with identifier-b is not nil

Comment (jbrophy17) [inline source/UberRidesTests/TokenManagerTests.swift]: Can we also add a test that verifies that the bool is saved to user defaults when a token is saved?

**Comment by jbrophy17** [comment]
@dayitv89 have you had a chance to look at my feedback? Would love to get this change merged in!

**Comment by gauravds** [comment]
@jbrophy17 I'm extremely sorry for the delay notice on this PR. Do we still need this changes or you have already implemented it? 
I can modify the above changes, should do?

**Comment by jbrophy17** [comment]
@gauravds no worries on the delay. I haven't had a chance to implement this, so if you could update it that would be great and I can get it merged :)

**Comment by edjiang** [comment]
Closing due to stale code and inactivity; feel free to reopen or track this issue via #101
